src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/DictionaryOfStringTValueConverter.cs

```diff
@@ -62,6 +62,7 @@ protected override void CreateCollection(ref ReadStack state)
                 {
                     string key = GetKeyName(enumerator.Current.Key, ref state, options);
                     writer.WritePropertyName(key);
+                    Debug.Assert(enumerator.Current.Value != null);
```

Comment:
I believe this assert is incorrect and will fail with this, so consider just using null-forgiving (`!`).
```C#
[Fact]
public static void WriteDictionaryNullValue()
{
    var input = new Dictionary<string, string> { { "a", null }, {"b", "foo" } };

    string json = JsonSerializer.Serialize(input);
}
```

Since this is calling the built-in `JsonConverter<string?>`, null is allowed and will work if passed in. The `StringConverter` overwrites the nullability, but I suppose it can't be expressed statically.
https://github.com/dotnet/runtime/blob/5d8f92435ebfeab67b41aba0bce12e1e8af3187e/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/StringConverter.cs#L14-L17

Is there something we can do here?

Reply:
I don't think we can do anything here. Since the converter here is declared as the abstract type, the compiler will resolve the signature that contains the `DisallowNull` annotation. So I think we should just use a `!`. I think we can't declare the converter variable type as `JsonConverter<string?>`, right?

Reply:
No, it has to remain generic. It just happens to be that all the Ts in the particular set of `JsonConverter<T>` work with null values (because our internal converters know how to deal with it).

I think, ideally, we will re-write parts of this to add special handling for null and honor the `DisallowNull` even internally.

Reply:
I think that would make sense. Internal usage of these APIs should overall honor the intent and what we tell our customers. It would be great if the usage of `!` was lower.